### PR TITLE
Improve `condition` support for Python and Rust

### PR DESCRIPTION
**Pull request by pokey**
- Adds `condition` support for Python ternaries
- Adds `condition` support for Python list comprehensions
- Adds `condition` support for Python match statements
- Adds `condition` support for Rust match statements
- Note that `"chuck"` includes the `if` keyword for match statements and list comprehensions
- Split from #831

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [x] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [x] I have not broken the cheatsheet

Co-authored-by: Michael Doronin <warrior2031@mail.ru>

### CODE REVIEW

**Comment by auscompgeek** [comment]
Seems reasonable. I presume `if` clauses in Python list comprehensions are treated similarly?

**Comment by pokey** [comment]
> Seems reasonable. I presume `if` clauses in Python list comprehensions are treated similarly?

Hmm I don't believe that they are.  Does make me wonder if it's what we want; feels a bit small for `"if state"`, as it's not a full statement.  Ie `"state"` should be a superset of `"if state"`, and I don't think `"state"` wants to get trapped this far down

Otoh I do sometimes find myself wanting to target the `if` in python comprehensions

cc/ @AndreasArvidsson wdyt

**Comment by AndreasArvidsson** [comment]
I guess it's similar to trying to target a ternary in other languages. I do agree that if statement isn't the best description for any of those, but it probably would be useful to be able to target them.

**Comment by pokey** [comment]
Discussed in meet-up:

- [ ] We'd like to switch this to `condition`
- [ ] For "change" it doesn't include the `if` keyword
- [ ] For `"chuck"` it _**does**_ include the `if` keyword

**Comment by pokey** [comment]
@purpleP fyi we are now using `"condition"` instead of `"if state"` here

File: src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  usePrePhraseSnapshot: true
+  action: {name: remove}
+initialState:
+  documentContents: "[aaa for aaa in bbb if ccc]"
+  selections:
+    - anchor: {line: 0, character: 1}
+      active: {line: 0, character: 1}
+  marks: {}
+finalState:
+  documentContents: "[aaa for aaa in bbb ]"
```
Comment (auscompgeek) [inline src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml:19]: Hmm, I was expecting the whitespace to disappear too. It looks like it does in the Rust implementation, wonder what's up here.

File: src/languages/rust.ts [hunk 2]
```diff
@@ -146,6 +146,7 @@ const nodeMatchers: Partial<
   ),
   string: ["raw_string_literal", "string_literal"],
   ifStatement: ["if_expression", "if_let_expression"],
+  condition: leadingMatcher(["*.match_pattern![condition]"], ["if"]),
```
Comment (auscompgeek) [inline src/languages/rust.ts:149]: I'm surprised this wasn't implemented in the first place. We should make sure this handles `if` and `while` conditions too.

Open question for how to handle `if let` and `while let`, though I imagine it should take the whole `let <match_pattern> = <value>`.

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml (hunk 1 above)
Comment (pokey) [reply src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml:19]: We generally consider this whitespace cleanup out of scope for scope types, as a) we have plans improve this stuff in the future in a more generic way, and b) an autoformatter can fix this one easily.  I'd be inclined to not bother fixing unless you feel strongly

File: src/languages/rust.ts (hunk 2 above)
Comment (pokey) [reply src/languages/rust.ts]: Done

File: src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml (hunk 1 above)
Comment (auscompgeek) [reply src/test/suite/fixtures/recorded/languages/python/ditchCondition.yml:19]: Fair enough. Just struck me as slightly odd; Rust condition not working as intended was the big one.

**Review by auscompgeek — APPROVED** [review]

**Pull request merged**